packages/loader/container-loader/src/audience.ts

```diff
@@ -3,29 +3,40 @@
  * Licensed under the MIT License.
  */
 
-import { EventEmitter } from "@fluid-internal/client-utils";
-import { IAudienceOwner } from "@fluidframework/container-definitions/internal";
+import { TypedEventEmitter } from "@fluid-internal/client-utils";
+import { IAudienceEvents, IAudienceOwner } from "@fluidframework/container-definitions/internal";
 import { assert } from "@fluidframework/core-utils";
 import { IClient } from "@fluidframework/protocol-definitions";
 
 /**
  * Audience represents all clients connected to the op stream.
  */
-export class Audience extends EventEmitter implements IAudienceOwner {
+export class Audience extends TypedEventEmitter<IAudienceEvents> implements IAudienceOwner {
 	private readonly members = new Map<string, IClient>();
+	private _currentClientId: string | undefined;
 
 	constructor() {
 		super();
 		// We are expecting this class to have many listeners, so we suppress noisy "MaxListenersExceededWarning" logging.
 		super.setMaxListeners(0);
 	}
 
-	public on(
-		event: "addMember" | "removeMember",
-		listener: (clientId: string, client: IClient) => void,
-	): this;
-	public on(event: string, listener: (...args: any[]) => void): this {
-		return super.on(event, listener);
+	public get currentClientId() {
+		return this._currentClientId;
+	}
+
+	public setCurrentClientId(clientId: string | undefined): void {
```

Comment:
Instead of having something outside call setCurrentClientId, I'd probably prefer passing in a callback getCurrentClientId() plus something to event when it changes - to avoid duplicating state storage (I note that a number of your comments are concerned with the risk of clientId getting out of sync between different "sources of truth", so maybe try to stick with only storing it in one place)

Reply:
Or if you prefer imperative to events, a notifyClientIdChanged() but still have audience go consult an external source of truth rather than storing it locally.

Reply:
I've moved all tracking to Audience in latest iteration - it's the only source of truth and there is no other source of the data (there is no more Container._clientId for example). We can do reverse (i.e. store it at container level), but I do not see much value - I want to have old clientId for events in Audience (to communicate old clientId when it changes), and we have to notify Audience when clientId changes (for it to raise events).

Reply:
I like that there's less duplication now, would love to see this go further though and eliminate the new property in audience.  The ConnectionManager/ConnectionStateHandler split is a bit confusing to me, but I suspect that one of those two would be the most sensible "one place to store it" (aside from reading it directly off the delta connection when online).

I see from your other comment that you like the "getMyself()" semantic (acknowledged that the implementation could benefit from some work) - did you consider just exposing a similar semantic from Audience rather than adding these new properties?

Reply:
1. **ConnectionManager** is only responsible for reconnection flow. It has no knowledge of semantics that other layers have.
2. **ConnectionStateHandler** manages such things as catchingUp -> Connected transition, i..e it waits for join op / leave. It also waits for leave op from previous connection, if that is required for correctness.
   - It is organized as a pipeline. I.e. Container does not work with one object instance / class, it works with abstraction, but under the cover it's organized as ConnectionStateHandler + ConnectionStateCatchup pipeline - former class is responsible for dealing with dealing with old connections and join op/leave waiting, the latter is responsible for waiting for particular sequence number (for "read" connections). There is room to split it even further into more atomic blocks (some time).
3. **Container** uses # 2, but also delays delivering of "connected" event / clientId change if we have connection established while loading. It also is responsible for "continuing" from clientId of the previous session if that is rehydration of state from stash (though latter is a bit split).

It's a pipeline (that's the order of how information moves in this pipeline, even though ownership is organized slightly differently), no single layer has "truth", each layer slightly modifies semantic of the truth.

There is no way to eliminate the property, it has to be somewhere. It can be in Audience, or it can be in Container, but it has to be somewhere. I want to remove all API surfaces that talk about clientId and leave just Audience (it will take some time to get there). I do not see any benefit to store that value in Container and use callbacks to read it by Audience.

Reply:
What do you mean by "similar semantic"? Similar to what? To AzureClient?
Yes, I've looked at it. I do not want to build complex thing (like all notion of users and group connections belonging to a user) - I think that belongs to a next layer. We can expose self as just a pair of clientId & IClient.
I'm not sure it solves the problem of how to tell user a story if it's "current" or "latest" info and when it changes - we have to do proper storytelling no matter what.

Reply:
added self() - take a look.

Reply:
> I want to remove all API surfaces that talk about clientId and leave just Audience (it will take some time to get there).

I'd instead push the other direction and have it only live on ConnectionStateHandler (because "clientId" is really something more like "connectionInstanceId", so one of these "connection managing" things seems the more appropriate owner.

> added self() - take a look.

Thanks, this is close to what I was thinking (added a few comments).

I'm using "semantic" here in a manner similar to how it's used for the "[semantic web](https://en.wikipedia.org/wiki/Semantic_Web)", in that it's a vocabulary for the customer to directly express their intent (of getting themselves from the audience) rather than an obscured fashion like `getMember(<client ID that happens to be my own>)`.  Amongst web people the usage of "semantic" broadened to more than just data tagging to include "APIs that convey customer intent".

Reply:
The question here is not what is the meaning of clientId, but when it changes (and who controls it). Each layer in pipeline has its own (consistent within that layer) answer to that question, but answers are vastly different across layers.

ConnectionStateHandler state and Container states differ. Container delays notification of connected / change of self on container load - this responsibility belongs to Container, not to ConnectionStateHandler.